lib/assert.js

```diff
@@ -314,21 +320,29 @@ function setEquiv(a, b, strict, memo) {
   if (a.size !== b.size)
     return false;
 
+  // This is a set of the entries in b which have been consumed in our pairwise
+  // comparison. Initialized lazily so sets which only have value types can
+  // skip an extra allocation.
+  let bEntriesUsed = null;
+
   for (const val1 of a) {
     // If the value doesn't exist in the second set by reference, and its an
     // object or an array we'll need to go hunting for something thats
     // deep-equal to it. Note that this is O(n^2) complexity, and will get
     // slower if large, very similar sets / maps are nested inside.
     // Unfortunately there's no real way around this.
-    if (!setHasSimilarElement(b, val1, strict, memo)) {
+    if (bEntriesUsed == null && typeof val1 === 'object')
```

Comment:
Maybe.. but in practice its very rare to mix types in set / map keys. And it still won't matter in strict mode because there's an early return for value types anyway.

Uhhhh but not in non-strict mode. Huh - looks like this also doesn't throw:

```javascript
assert.deepEqual(new Set([3, '3']), new Set([3, 4]));
```

*fixes*